pkg/provider/ngrok/client.go

```diff
@@ -0,0 +1,237 @@
+	errVaultSecretDoesNotExist = errors.New("vault secret does not exist")
+	errCannotPushNilSecret     = errors.New("cannot push nil secret")
+)
+
+type VaultClient interface {
+	Create(context.Context, *ngrok.VaultCreate) (*ngrok.Vault, error)
+	List(*ngrok.Paging) ngrok.Iter[*ngrok.Vault]
+}
+
+type SecretsClient interface {
+	Create(context.Context, *ngrok.SecretCreate) (*ngrok.Secret, error)
+	Update(context.Context, *ngrok.SecretUpdate) (*ngrok.Secret, error)
+	Delete(context.Context, string) error
+	List(*ngrok.Paging) ngrok.Iter[*ngrok.Secret]
+}
+
+type client struct {
+	vaultClient   VaultClient
+	secretsClient SecretsClient
+	vaultName     string
+}
+
+func (c *client) PushSecret(ctx context.Context, secret *corev1.Secret, data esv1.PushSecretData) error {
+	if secret == nil {
+		return errCannotPushNilSecret
+	}
+
+	// First, get the vault by name. If it doesn't exist, create it.
+	vault, err := c.getVaultByName(ctx, c.vaultName)
+	if err != nil {
+		if !errors.Is(err, errVaultDoesNotExist) {
+			return err
+		}
+
+		// Create the vault if it does not exist
+		vault, err = c.vaultClient.Create(ctx, &ngrok.VaultCreate{
+			Name:        c.vaultName,
+			Description: defaultDescription,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create vault: %w", err)
+		}
+	}
+
+	// Prepare the secret data for pushing
+	value, err := json.Marshal(secret.Data)
+	if err != nil {
+		return fmt.Errorf("json.Marshal failed with error: %w", err)
+	}
+
+	// If key is specified, get the value from the secret data
+	if data.GetSecretKey() != "" {
+		var ok bool
+		value, ok = secret.Data[data.GetSecretKey()]
+		if !ok {
+			return fmt.Errorf("key %s not found in secret", data.GetSecretKey())
+		}
+	}
+
+	// Check if the secret already exists in the vault
+	existingSecret, err := c.getSecretByVaultAndName(ctx, *vault, data.GetRemoteKey())
+	if err != nil {
+		if !errors.Is(err, errVaultSecretDoesNotExist) {
+			return fmt.Errorf("failed to get secret: %w", err)
+		}
+
+		// If the secret does not exist, create it
+		_, err = c.secretsClient.Create(ctx, &ngrok.SecretCreate{
+			VaultID:     vault.ID,
+			Name:        data.GetRemoteKey(),
+			Value:       string(value),
+			Metadata:    data.GetMetadata().String(),
+			Description: defaultDescription,
+		})
+		return err
+	}
+
+	// If the secret exists, update it
+	_, err = c.secretsClient.Update(ctx, &ngrok.SecretUpdate{
+		ID:          existingSecret.ID,
+		Value:       ptr.To(string(value)),
+		Metadata:    ptr.To(data.GetMetadata().String()),
+		Description: ptr.To(string(defaultDescription)),
+	})
+	return err
+}
+
+func (c *client) SecretExists(ctx context.Context, ref esv1.PushSecretRemoteRef) (bool, error) {
+	// Implementation for checking if a secret exists in ngrok
+	secret, err := c.getSecretByVaultNameAndSecretName(ctx, c.vaultName, ref.GetRemoteKey())
+	if err != nil {
+		if errors.Is(err, errVaultDoesNotExist) || errors.Is(err, errVaultSecretDoesNotExist) {
```

Comment:
same nit as above for moving out of if err for nesting reduction

Reply:
Fixed in https://github.com/external-secrets/external-secrets/pull/5160/commits/b603c6166430b01b58605f0048ab1cf33251bcfe